### PR TITLE
Implement kubertas daemonset and service.

### PR DESCRIPTION
**Pull request by kishen-v**
Implement kubertas daemonset and service.

### CODE REVIEW

**Comment by kishen-v** [comment]
cc @Karthik-K-N

**Comment by mkumatag** [comment]
@Rajalakshmi-Girish can you please enable prow for this repo like web hook etc?

File: cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	serviceConfig, err := utils.ReadConfig(cfgFilePath)
+	if err != nil {
+		klog.Errorf("error while reading config. Error: %v", err)
+		return
+	}
+
+	// Check if the servicelog.db file is present in the path set in config file.
+	if _, err := os.Stat(serviceConfig.ServicelogPath); err != nil {
+		klog.Errorf("Error finding the servicelog.db file. %v", err)
+		return
```
Comment (Karthik-K-N) [inline cmd/main.go]: how about using klog.Fatalf(), instead of Errorf and return

File: cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+	// Capture any OS signals to perform graceful exit.
+	sigchan := make(chan os.Signal, 1)
+	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		sig := <-sigchan
+		done <- struct{}{}
+		klog.Infof("Captured %s, Shutting down KubeRTAS", sig.String())
+		wg.Done()
+		os.Exit(0)
```
Comment (Karthik-K-N) [inline cmd/main.go:85]: wont it automatically exit the program after wg.Done(), is it used to be on safer side?

File: dbreader/dbreader.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+	"k8s.io/klog/v2"
+
+	// SQLite driver.
+	_ "rsc.io/sqlite"
+)
+
+// SQL Queries to retrieve entries from the servicelog.db file on the host.
+const (
+	logFilter    = "SELECT ID,SEVERITY,DESCRIPTION FROM EVENTS WHERE ID > %v AND SEVERITY > %v"
```
Comment (Karthik-K-N) [inline dbreader/dbreader.go]: may be a var declaration instead of const, I think const should never change

File: dbreader/dbreader.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	return &ServicelogParser{
+		dbPath:   dbFilePath,
+		severity: eventSeverity,
+	}
+}
+
+// ReadServicelogDB accesses the servicelog.db and retrieves the rows, based on
+// the severity threshold configured. The retrieved entries are logged and published as
+// an event to the kubernetes API server.
+func (Reader *ServicelogParser) ReadServicelogDB(notifier *knode.K8sNotifier) error {
```
Comment (Karthik-K-N) [inline dbreader/dbreader.go]:
```suggestion
func (reader *ServicelogParser) ReadServicelogDB(notifier *knode.K8sNotifier) error {
```

File: knode/knode.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/record"
+
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
```
Comment (Karthik-K-N) [inline knode/knode.go]: may be all k8s imports can be clubbed together

File: knode/knode.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+// server.
+func (k *K8sNotifier) InitializeK8sNotifier() error {
+	var err error
+	if k.inCluster {
+		k.config, err = rest.InClusterConfig()
+		if err != nil {
+			return err
+		}
+		k.nodeName = os.Getenv("NODE_NAME")
+		if err != nil {
```
Comment (Karthik-K-N) [inline knode/knode.go]: I think it should be k.nodeName == ""

File: cmd/main.go (hunk 2 above)
Comment (kishen-v) [reply cmd/main.go:85]: @Karthik-K-N, It will automatically exit. I've kept it nevertheless.

File: cmd/main.go (hunk 1 above)
Comment (kishen-v) [reply cmd/main.go]: Will incorporate the suggestion.

File: knode/knode.go (hunk 6 above)
Comment (kishen-v) [reply knode/knode.go]: You are right.. Thanks for pointing out. 😃

File: knode/knode.go (hunk 5 above)
Comment (kishen-v) [reply knode/knode.go]: Done.

File: dbreader/dbreader.go (hunk 4 above)
Comment (kishen-v) [reply dbreader/dbreader.go]: Made the required changes.

File: dbreader/dbreader.go (hunk 3 above)
Comment (kishen-v) [reply dbreader/dbreader.go]: Moved the strings to their respective queries.. as they're not reused anywhere. Removed the `const` section altogether.

File: dbreader/dbreader.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		for rows.Next() {
+			err = rows.Scan(&config.Id, &config.Sev, &config.Desc)
+			if err != nil {
+				return fmt.Errorf("cannot read rows from DB %v", err)
+			}
+			if logEntry, err := json.MarshalIndent(config, "", " "); err != nil {
+				return fmt.Errorf("cannot marshal data %v", err)
+			} else {
+				klog.Infof("Retrieved RTAS event:\n%v", string(logEntry))
+			}
```
Comment (Karthik-K-N) [inline dbreader/dbreader.go]: No need of else block, It can directly go as next statement as we are returning in case of if

File: knode/knode.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	var err error
+	if k.config, err = config.GetConfig(); err != nil {
+		klog.Fatalf("cannot retrieve rest config for kubernetes. %v ", err)
+	}
+	// Proceed to fetch the NODE_NAME set through the downward API.
+	k.nodeName = os.Getenv("NODE_NAME")
+	// If the node name is still unset, obtain the hostname as it may be running in standalone
+	// mode.
+	if k.nodeName == "" {
+		k.nodeName, err = os.Hostname()
```
Comment (Karthik-K-N) [inline knode/knode.go]: may be we can make it as mandatory to set NODE_NAME environmental variable and avoid fetching hostname what if user dont know how node name is fetched and tries running from outside cluster

**Review by Karthik-K-N — COMMENTED** [review]
small comment/dount otherwise LGTM

File: knode/knode.go (hunk 8 above)
Comment (kishen-v) [reply knode/knode.go]: By design in standalone mode, the application must run on the node from which the servicelogs needs to be processed + the dependency from Kubecofig.. In my opinion, obtaining the node name from the system removes chances where:
 1. The node with the set NODE_NAME environment variable may not exist. 
 2. Misconfiguration. 
 3. Typos(?). 

It is most likely that the hostname is not different from the kubernetes node-name.. Or maybe I'm missing the point why it'd be better to do away with the part where we determine hostname..

File: dbreader/dbreader.go (hunk 7 above)
Comment (kishen-v) [reply dbreader/dbreader.go]: Addressed..

**Comment by Karthik-K-N** [comment]
LGTM

**Review by mkumatag — CHANGES_REQUESTED** [review]
@kishen-v I remember Mahesh had a concern about calling the project `kube-rtas`, can you confirm with him once?

File: knode/knode.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+var Kubeconfig *string
+
+type K8sNotifier struct {
```
Comment (mkumatag) [inline knode/knode.go]: Any specific reason for naming this as `K8sNotifier` or can we just keep it simple like `Notifier`?

File: knode/knode.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	recorder record.EventRecorder
+}
+
+func NewK8sNotifier() *K8sNotifier {
+	return &K8sNotifier{}
+}
+
+// InitializeK8sNotifier sets up the event recorder to send events to the Kube API
+// server.
+func (k *K8sNotifier) InitializeK8sNotifier() error {
```
Comment (mkumatag) [inline knode/knode.go]:
```suggestion
func (k *K8sNotifier) Initialize() error {
```

File: knode/knode.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	var err error
+	if k.config, err = config.GetConfig(); err != nil {
+		klog.Fatalf("cannot retrieve rest config for kubernetes. %v ", err)
```
Comment (mkumatag) [inline knode/knode.go]: you should return error if the function already has a return statement for that

File: knode/knode.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	// Proceed to fetch the NODE_NAME set through the downward API.
+	k.nodeName = os.Getenv("NODE_NAME")
+	// If the node name is still unset, obtain the hostname as it may be running in standalone
+	// mode.
+	if k.nodeName == "" {
```
Comment (mkumatag) [inline knode/knode.go]: let us add some warning/info message which gives heads up to the user that code is executing in the standalone mode

File: knode/knode.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+			return errors.New("hostname is unavailable, node events cannot be reported")
+		}
+	}
+
+	client := kubernetes.NewForConfigOrDie(k.config).CoreV1()
+	k.recorder = getEventRecorder(client, "", k.nodeName, "RTASNotifier")
+	return nil
+}
+
+func getEventRecorder(c typedcorev1.CoreV1Interface, namespace, nodeName, source string) record.EventRecorder {
```
Comment (mkumatag) [inline knode/knode.go]: if namespace is not passed from the calling function then lets remove it

File: knode/knode.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+	return recorder
+}
+
+// NotifyAPIServer posts events to the Kubernetes API server.
+func (k *K8sNotifier) NotifyAPIServer(message string) {
+	k.recorder.Event(&v1.ObjectReference{
+		Kind: "Node",
+		Name: k.nodeName,
+		UID:  types.UID(k.nodeName),
+	}, v1.EventTypeWarning, "PlatformEvent", message)
```
Comment (mkumatag) [inline knode/knode.go]: why event is always `v1.EventTypeWarning`? can we set the severity type based on the type of events we are getting from the DB?

File: dbreader/dbreader.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+	return &ServicelogParser{
+		dbPath:   dbFilePath,
+		severity: eventSeverity,
+	}
+}
+
+// ReadServicelogDB accesses the servicelog.db and retrieves the rows, based on
+// the severity threshold configured. The retrieved entries are logged and published as
+// an event to the kubernetes API server.
+func (reader *ServicelogParser) ReadServicelogDB(notifier *knode.K8sNotifier) error {
```
Comment (mkumatag) [inline dbreader/dbreader.go]: this function name sounded like only reads the event and also the package name is dbreader! so I see a scope for making the code better here.

File: dbreader/dbreader.go [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+	if err != nil {
+		return fmt.Errorf("error while opening DB. %v", err)
+	}
+	// Estimate the number of rows that are needed to be processed
+	if err = db.QueryRow(fmt.Sprintf("SELECT COUNT(*) FROM EVENTS WHERE ID > %v",
+		reader.currLogCount)).Scan(&reader.currLogCount); err != nil {
+		return fmt.Errorf("error while retrieving number of rows %v", err)
+	}
+	// Check if the number of rows have increased in DB since last read and process pending logs, if any.
+	if reader.currLogCount-reader.prevLogCount > 0 {
```
Comment (mkumatag) [inline dbreader/dbreader.go]: instead you can check for length 0 and return directly from here to have early return

File: dbreader/dbreader.go [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+		klog.V(1).Infof("Processing %d logs", reader.currLogCount-reader.prevLogCount)
```
Comment (mkumatag) [inline dbreader/dbreader.go]: why is the logging level set to 1? can you check the default log level and set higher if you want to see this as debug or log level set greater than the default

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: knode/knode.go (hunk 14 above)
Comment (kishen-v) [reply knode/knode.go]: Agree, we can cascade the severity from the RTAS log, I was trying to keep it inline with the pre-defined types of events in K8s. 
Should it be okay to deviate from the existing type? 
https://github.com/kubernetes/api/blob/ba9d27cc9c7de0882d39035d8ae3b18d1434b636/core/v1/types.go#L6159-L6165

Comment (mkumatag) [reply knode/knode.go]: na, what I really meant is - we can map the events from RTAS logs to k8s event types

Comment (kishen-v) [reply knode/knode.go]: Got your point.. But I still have some confusion.

The defined k8s event types are only Warning and Normal. While there may be events from other components with the existing type, the RTAS events are of 7 different levels - Debug, Info, Event, Warning, Error_local, Error and Fatal. While we can Map Warning to warning and Debug and Info to Normal events, the types Error_local, Error and Fatal don't have a corresponding severity type defined in core-types..

My doubt is whether it'd be okay to have new types of severity appearing under events( Like Error, Fatal, etc.), while the available Kubernetes defined types of events are only Normal and Warning..

File: dbreader/dbreader.go (hunk 17 above)
Comment (kishen-v) [reply dbreader/dbreader.go]: The default log level is 0. In this case, we have only have 2 supported levels of logging, 0(Info) and 1(Debug). The notion was to have the line printed only when the verbosity is set to 1.

File: knode/knode.go (hunk 14 above)
Comment (mkumatag) [reply knode/knode.go]: In that case let's map to whatever levels we have, like serious events as warning and not so serious issue as Info

File: knode/knode.go (hunk 9 above)
Comment (kishen-v) [reply knode/knode.go]: We can keep it simply  -`Notifier`.

File: knode/knode.go (hunk 11 above)
Comment (kishen-v) [reply knode/knode.go]: Made changes to return the error and log from the caller, if needed.

File: knode/knode.go (hunk 12 above)
Comment (kishen-v) [reply knode/knode.go]: Added required info logs.

**Review by mkumatag — APPROVED** [review]
/lgtm

**Pull request merged**